include/cocaine/framework/receiver.hpp

```diff
@@ -0,0 +1,324 @@
+    typedef value_type result_type;
+
+    static
+    result_type
+    convert(variant_type& value) {
+        return boost::apply_visitor(visitor_t(), value);
+    }
+
+private:
+    struct visitor_t : public boost::static_visitor<result_type> {
+        result_type operator()(value_type& value) const {
+            return value;
+        }
+
+        result_type operator()(error_type& error) const {
+            throw cocaine_error(error);
+        }
+    };
+};
+
+template<class T>
+struct receiver_traits<io::streaming_tag<T>> {
+private:
+    typedef typename variant_of<io::streaming_tag<T>>::type variant_type;
+
+    typedef typename boost::mpl::at<typename variant_type::types, boost::mpl::int_<0>>::type value_type;
+    typedef typename boost::mpl::at<typename variant_type::types, boost::mpl::int_<1>>::type error_type;
+    typedef typename boost::mpl::at<typename variant_type::types, boost::mpl::int_<2>>::type choke_type;
+
+public:
+    typedef boost::optional<value_type> result_type;
+
+    static
+    result_type
+    convert(variant_type& value) {
+        return boost::apply_visitor(visitor_t(), value);
+    }
+
+private:
+    struct visitor_t : public boost::static_visitor<result_type> {
+        result_type operator()(value_type& value) const {
+            return boost::make_optional(value);
+        }
+
+        result_type operator()(error_type& error) const {
+            throw cocaine_error(error);
+        }
+
+        result_type operator()(choke_type&) const {
+            return boost::none;
+        }
+    };
+};
+
+template<class T>
+class terminator {
+    typedef std::function<bool()> function_type;
+
+public:
+    static std::vector<function_type> generate() {
+        std::vector<function_type> result;
+        boost::mpl::for_each<typename io::protocol<T>::messages>(terminator<T>(result));
+        return result;
+    }
+
+    template<class U>
+    void operator()(const U&) {
+        unpackers.emplace_back(&terminator::unpacker<U>::unpack);
+    }
+
+private:
+    std::vector<function_type>& unpackers;
+
+    terminator(std::vector<function_type>& unpackers) :
+        unpackers(unpackers)
+    {}
+
+public:
+    template<typename U>
+    struct unpacker {
+        static
+        bool
+        unpack() {
+            return std::is_same<typename io::event_traits<U>::dispatch_type, void>::value;
+        }
+    };
+};
+
+template<class T, class Session>
+class receiver {
+    typedef typename variant_of<T>::type variant_type;
+    typedef typename variant_type::types variant_typelist;
+
+    typedef std::function<variant_type(const msgpack::object&)> unpacker_type;
+    typedef std::function<bool()> terminator_type;
+
+    static const std::vector<unpacker_type> visitors;
+    static const std::vector<terminator_type> terminators;
+
+    std::shared_ptr<basic_receiver_t<Session>> d;
+
+public:
+    receiver(std::shared_ptr<basic_receiver_t<Session>> d) :
+        d(d)
+    {}
+
+    // Intentionally deleted.
```

Comment:
Yep (for now), because of streaming receiver. I think I'll make it deleted for the common case, but instead of will create additional receiver specialization for `streamig_tag<T>`.